fairlearn/metrics/metrics_engine.py

```diff
@@ -0,0 +1,105 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+
+from . import DisparityResult, MetricResult
+
+_MESSAGE_NON_BINARY = "Array {0} contains values other than 0 and 1"
+
+
+def metric_by_groups(metric_function, y_true, y_pred, group_data, sample_weight=None):
+    """ Applies a metric to each subgroup of a set of data
+
+    :param metric_function
+    :type Function with signature (y_true, y_pred, sample_weight=None) which returns a scalar
+
+    :param y_true
+    :type Array of actual results (must be 0 or 1)
+
+    :param y_pred
+    :type Array of predicted results (must be 0 or 1)
+
+    :param group_data
+    :type Array indicating the group to which each result belongs
+
+    :param sample_weight
+    :type Array of weights to apply to each result
+    """
+    _check_binary(y_true, "y_true")
+    _check_binary(y_pred, "y_pred")
+    result = MetricResult()
+
+    groups = np.unique(group_data)
+
+    result.metric = metric_function(y_true, y_pred, sample_weight=sample_weight)
+
+    # The slicing we use requires Numpy arrays
+    y_a = np.array(y_true)
+    y_p = np.array(y_pred)
+    s_w = None
+    if sample_weight is not None:
+        s_w = np.array(sample_weight)
+    for group in groups:
+        group_indices = [i for i, elem in enumerate(group_data) if elem == group]
```

Comment:
this looks inefficient. you should be able to do something like group_indices = (group_data==elem), which is handled in C rather than the above code that looks like the vector creation is handled by python.

Reply:
(i mean it's handled in C if group_data is ndarray)

Reply:
Converted; I had suspected there might be a quicker way